azure/services/resource.ftl

```diff
@@ -159,6 +159,11 @@
         [#return ""]
     [/#if]
 
+    [#-- Check if its an undeployed pseudo stack --]
+    [#if (profile.type == "pseudo") && !(name?has_content)]
+        [#return getExistingReference(id, attributeType)]
+    [/#if]
+
```

Comment:
I don't understand why this isn't redundant with the L145 test "! isPartOfCurrentDeploymentUnit(id)", shouldn't all pseudo's fail that in any case?

Reply:
This is a pretty complex one that touches on a lot of things. It's a good question that requires a fairly lengthy explanation, ill see how I go.

Within a provider, the Services that are defined will each add Resources. For Azure Resource Manager templates these are direct references to Azure resources. Each separate resource is a unique type.

In Hamlet resource types define their outputs up front - Both Azure and AWS have a maximum number of outputs you can define per template. Whilst that number looks high, it quickly adds up when you have a large number of outputs and resources per template. If a Policy resource on for a Virtual Network used 5 outputs, then a project requires a dozen or so Policy resources, they get eaten up fast. So to avoid this, resource types only define the essential outputs that are absolutely required in order to pass information on to another part of the engine.

During template generation for a given deployment unit, the state routine (state.ftl macros) adds the resources that collectively make up it's component type to the Solution. As they are added, the resource types are evaluated, and the necessary outputs for that resource type are added to the StackOutputsObject - a data structure that will be converted into the template outputs when serialised, but for now its just a JSON object.

The function `isPartOfCurrentDeploymentUnit()` makes a comparison between:
a) the deployment-unit associated with the resourceId passed to the function in the StackOutputsObject, and;
b) the deployment unit that the engine knows itself to be currently working with.
If they are different, then the resource defined is not in the current Deployment Unit.

Where this all breaks down with Pseudo resources is that they are the exception to the mapping between Azure Resource Manager resources and Hamlet Resources - they don't necessarily exist in ARM, but conceptually in Hamlet they are still resources. Often they are still known to be resources by the provider (if you create them through the GUI, they may be exposed as resources) but for whatever reason they haven't been exposed through ARM templates. So we instead define a pseudo-resource type for them, which allows us to leverage the existing resource loading behaviour. 

**However,**

when it comes to stack outputs, we cannot include the pseudo stack outputs into the template (they might not get deployed until after the template.json is deployed, most commonly they find themselves deployed in epilogue scripts), so they do not get their outputs added to the StackOutputsObject. Because of this behaviour the function you have specified will not correctly identify if pseudo resource types are refined in the current deployment unit and they need to be handled separately.

Reply:
Thanks